src/Formats/SchemaInferenceUtils.cpp

```diff
@@ -697,55 +705,87 @@ namespace
             return false;
 
-        ReadBufferFromString buf(field);
-        Float64 tmp_float;
         /// Check if it's just a number, and if so, don't try to infer Date from it,
         /// because we can interpret this number as a Date (for example 20000101 will be 2000-01-01)
         /// and it will lead to inferring Date instead of simple Int64/UInt64 in some cases.
-        if (tryReadFloatText(tmp_float, buf) && buf.eof())
+        if (std::all_of(field.begin(), field.end(), isNumericASCII))
             return false;
 
-        buf.seek(0, SEEK_SET); /// Return position to the beginning
-
+        ReadBufferFromString buf(field);
         DayNum tmp;
-        return tryReadDateText(tmp, buf) && buf.eof();
+        return tryReadDateText(tmp, buf, DateLUT::instance(), /*allowed_delimiters=*/"-/:") && buf.eof();
     }
 
-    bool tryInferDateTime(std::string_view field, const FormatSettings & settings)
+    DataTypePtr tryInferDateTimeOrDateTime64(std::string_view field, const FormatSettings & settings)
     {
-        if (field.empty())
-            return false;
+        /// Don't try to infer DateTime if string is too long.
+        /// It's difficult to say what is the real maximum length of
+        /// DateTime we can parse using BestEffort approach.
+        /// 50 symbols is more or less valid limit for date times that makes sense.
+        if (field.empty() || field.size() > 50)
+            return nullptr;
+
+        /// Check that we have at least one digit, don't infer datetime form strings like "Apr"/"May"/etc.
+        if (!std::any_of(field.begin(), field.end(), isNumericASCII))
+            return nullptr;
 
-        ReadBufferFromString buf(field);
-        Float64 tmp_float;
         /// Check if it's just a number, and if so, don't try to infer DateTime from it,
         /// because we can interpret this number as a timestamp and it will lead to
-        /// inferring DateTime instead of simple Int64/Float64 in some cases.
+        /// inferring DateTime instead of simple Int64 in some cases.
+        if (std::all_of(field.begin(), field.end(), isNumericASCII))
+            return nullptr;
+
+        ReadBufferFromString buf(field);
+        Float64 tmp_float;
+        /// Check if it's a float value, and if so, don't try to infer DateTime from it,
+        /// because it will lead to inferring DateTime instead of simple Float64 in some cases.
         if (tryReadFloatText(tmp_float, buf) && buf.eof())
-            return false;
+            return nullptr;
+
+        buf.seek(0, SEEK_SET); /// Return position to the beginning
+        if (!settings.try_infer_datetimes_only_datetime64)
+        {
+            time_t tmp;
+            switch (settings.date_time_input_format)
+            {
+                case FormatSettings::DateTimeInputFormat::Basic:
+                    if (tryReadDateTimeText(tmp, buf, DateLUT::instance(), /*allowed_date_delimiters=*/"-/:", /*allowed_time_delimiters=*/":") && buf.eof())
```

Comment:
Another common delimiter is `.`. Especially for dates and it might be used with fractional seconds separated by another dot. Should we also have it? (here and/or below)

Reply:
I don't want to allow `.` delimiter because of cases like `{"version" : "1.1.10"}` that can be parsed as `2010-01-01 00:00:00`.

Reply:
And I forgot to add a test for this case and in `best_effort` we don't check allowed delimieters in one place: https://github.com/ClickHouse/ClickHouse/pull/68452

Reply:
But, it was more or less expected behavior for people, and it is backward incompatible change.

https://fiddle.clickhouse.com/896ef97d-5898-4aff-97d1-06113fc3d7c9